src/stream/src/executor/backfill/cdc/cdc_backfill.rs

```diff
@@ -483,6 +521,70 @@ impl<S: StateStore> CdcBackfillExecutor<S> {
     }
 }
 
+fn mapping_message(
+    msg: Message,
+    upstream_indices: &[usize],
+    offset_parse_func: &CdcOffsetParseFunc,
+    last_cdc_offset: &Option<CdcOffset>,
+    upstream_table_name: &str,
+    upstream_table_id: u32,
+) -> Option<Message> {
+    match msg {
+        Message::Barrier(_) => Some(msg),
+        Message::Watermark(watermark) => watermark
+            .transform_with_indices(upstream_indices)
+            .map(Message::Watermark),
+        Message::Chunk(chunk) => {
+            let mark_result = mark_upstream_chunk(chunk, offset_parse_func, last_cdc_offset);
+            match mark_result {
+                Ok(chunk) => Some(Message::Chunk(mapping_chunk(chunk, upstream_indices))),
+                Err(e) => {
+                    tracing::error!("failed to mark upstream chunk: {:?}", e.as_report());
+                    GLOBAL_ERROR_METRICS.cdc_source_error.report([
+                        upstream_table_name.into(),
+                        upstream_table_id.to_string(),
+                        e.to_report_string(),
+                    ]);
+                    None
+                }
+            }
+        }
+    }
+}
+
+/// filter out upstream events occurred before `last_cdc_offset`
+fn mark_upstream_chunk(
```

Comment:
why do we need this? I thought RW just reads from the latest CDC message?

Reply:
That's how we define latest. The table should only consume changes after it has been created.
Without this filter, it could ingest changes before the table created.

Reply:
> Without this filter, it could ingest changes occurred before the table created.

Can you elaborate more? Since we're not backfilling here, why there's still be historical data?

Reply:
I didn't dig into the root cause but it does occurred in the test. I think the reason could be we don't acquire lock on upstream db so that the starting point of binlog is not deterministic.

Reply:
Is there a way to fix this? In my opinion, if we can guarantee that the source job only emits events starting from the latest snapshot, then we don't need the `Backfill` executor at all. This would make it more consistent with other backfill sources like Kafka.

Reply:
>Is there a way to fix this?

The source is undeterministic w/o lock, so we read the current binlog offset as a low watermark and filter out events < low watermark (`last_cdc_offset` in code).

>`we don't need the Backfill executor at all. This would make it more consistent with other backfill sources like Kafka.`

Hmm, I don't get it. Why need to be consistent with other backfill sources? It is unnecessary to change stream plan for `snapshot=false`, and the upstream message is in `payload jsonb` schema, we need to parse the payload into table columns in the CdcBackfill.

Reply:
Got it, thx. It's just a matter of implementation so never mind.

> and the upstream message is in `payload jsonb` schema, we need to parse the payload into table columns in the CdcBackfill

Oh, didn't realize that. Is it exactly done in the `Backfill` executor, or a `Project` after that?

Reply:
> so that the starting point of binlog is not deterministic

Actually I think there's no accurate definition of what "latest offset" is due to network latency and buffering. So IMO there's no need to be such serious. 🤣

Reply:
The schema transform is done in the cdc backfill executor right now.

Reply:
> > so that the starting point of binlog is not deterministic

> Actually I think there's no accurate definition of what "latest offset" is due to network latency and buffering. So IMO there's no need to be such serious. 🤣

+1. I believe it is not worth doing so.

For testing purpose, you may try [`CHECKPOINT`](https://www.postgresql.org/docs/current/sql-checkpoint.html) command before connecting RW to it with `snapshot=false`. I think it will flush and cut the WAL.

Reply:
Ok, we can revisit whether it is needed to filter the upstream events with binlog low watermark in future.